modules/datastore/src/DataDictionary/AlterTableQuery/Builder.php

```diff
@@ -0,0 +1,218 @@
+<?php
+
+namespace Drupal\datastore\DataDictionary\AlterTableQuery;
+
+use Drupal\common\Storage\DatabaseConnectionFactoryInterface;
+use Drupal\datastore\DataDictionary\AlterTableQueryBuilderInterface;
+use Drupal\datastore\DataDictionary\AlterTableQueryInterface;
+
+use PDLT\ConverterInterface;
+use RootedData\RootedJsonData;
+
+/**
+ * Alter table query builder.
+ */
+class Builder implements AlterTableQueryBuilderInterface {
+
+  /**
+   * Database connection factory.
+   *
+   * @var \Drupal\common\Storage\DatabaseConnectionFactoryInterface
+   */
+  protected DatabaseConnectionFactoryInterface $databaseConnectionFactory;
+
+  /**
+   * Internal to SQL date format converter.
+   *
+   * @var \PDLT\ConverterInterface
+   */
+  protected ConverterInterface $dateFormatConverter;
+
+  /**
+   * Alter query class name.
+   *
+   * @var string
+   */
+  protected string $queryClass;
+
+  /**
+   * Alter query table name.
+   *
+   * @var string
+   */
+  protected string $table;
+
+  /**
+   * Alter query field names, types, and formats.
+   *
+   * @var array[]
+   *
+   * Example format:
+   * @code
+   * [
+   *   [
+   *     'name' => 'some_date_field',
+   *     'title' => 'Some Date Field', // optional
+   *     'type' => 'date',
+   *     'format' => '%Y-%m-%d' // optional
+   *   ],
+   * ]
+   * @endcode
+   */
+  protected array $fields = [];
+
+  /**
+   * Alter query index names, types, and formats.
+   *
+   * @var array[]
+   *
+   * Example format:
+   * @code
+   * [
+   *   [
+   *     'name' => 'some_date_field', // optional
+   *     'type' => 'fulltext', // optional
+   *     'fields' => [
+   *       [
+   *         'name' => 'field_a',
+   *         'length' => 10 // optional
+   *       ],
+   *       [
+   *         'name' => 'field_b',
+   *         'length' => 8 // optional
+   *       ],
+   *     ]
+   *   ]
+   * ]
+   * @endcode
+   */
+  protected array $indexes = [];
+
+  /**
+   * Create an alter table query factory.
+   */
+  public function __construct(
+    DatabaseConnectionFactoryInterface $database_connection_factory,
+    ConverterInterface $date_format_converter,
+    string $query_class
```

Comment:
I'm not sure about this pattern... wouldn't it make more sense to do this through normal dependency injection and type hint it to the `AlterTableQueryInterface`?